Add Github pull request support for Leo

bbondy opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/35492

<img width="1467" alt="Screenshot 2024-02-23 at 2 11 55 PM" src="https://github.com/brave/brave-core/assets/831718/7c9f7d31-0c2a-4510-aab9-60df37422e73">


This obtains the github patch file when the user is on a URL of the form github.com/[org]/[repo]/pull{/...} and uses that patch file as the content for Leo to consume.

Added the needs-security-review label and thypon because it adds a network request to github.com for the patch file when on a github.com pull request.

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-arm64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-x64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

thypon commented:
Q: does this work with PR in private repos?

bbondy commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -104,6 +127,33 @@ class PageContentFetcher {
+                                      weak_ptr_factory_.GetWeakPtr(),
+                                      std::move(callback), std::move(loader));
+    loader_ptr->DownloadToString(url_loader_factory_.get(),
+                                 std::move(on_response), 2 * 1024 * 1024);
```
that's just a limit on the response size and it gets truncated after that. It's using what transcripts are already using in that same file currently.

bbondy commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -104,6 +127,33 @@ class PageContentFetcher {
+    request->url = patch_url;
+    request->load_flags = net::LOAD_DO_NOT_SAVE_COOKIES;
+    request->credentials_mode = network::mojom::CredentialsMode::kInclude;
+    request->request_initiator = absl::nullopt;
```
this was done because of a network trace for a top level load of .patch, it matches the value for that when I set it like this.

bbondy replied:
I can probably remove that line. I’ll check.  I didn’t know why at first when I was developing it, but a top level load would get a private repo patch but my code was not. So I did a network-log to see what the differences were. And I set things so it appeared as a top level load. I think this is the only bit that actually matters for that `request->credentials_mode = network::mojom::CredentialsMode::kInclude;`

diracdeltas replied:
``` 
  // Failure to provide a correct, verified |request_initiator| may
  // lead to bypasses of CORS, ORB, SameSite cookies and other HTTP security
  // features.  An untrustworthy process (e.g. a renderer or a utility process)
  // should not be able to trigger or influence requests with a absl::nullopt
  // |request_initiator|).
  ```
  
from https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/mojom/url_request.mojom;l=161-169?q=url_request.mojom&ss=chromium it seems most appropriate to set the request_initiator to the PR URL. 

it seems `absl::nullopt` should really be used only for user-initiated requests like direct navigation in the omnibox or clicking a bookmark.

i don't see any security issues in the way it is now, so this is non-blocking.

bbondy replied:
I originally used `absl::nullopt` when developing the feature because it was retrieving the patch file via a top level load and not from my code. And I noticed not setting it is different from setting it to `absl::nullopt` in the network log, so I was trying to make it appear as top level.  But that wasn't needed. I'll leave it as is though with the line removed now because our other usages do the same.

darkdh commented on components/ai_chat/content/browser/page_content_fetcher.h:
```diff
@@ -22,7 +27,9 @@ using FetchPageContentCallback =
-                      FetchPageContentCallback callback);
+                      FetchPageContentCallback callback,
+                      scoped_refptr<network::SharedURLLoaderFactory>
+                          url_loader_factory = nullptr);
```
url_loader_factory_for_test

darkdh commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -348,11 +422,32 @@ std::string ExtractPdfContent(const ui::AXNode* pdf_root) {
+    return std::nullopt;
+  }
+  std::string patch_url_str = url.GetWithEmptyPath().spec() + path_parts[0] +
+                              "/" + path_parts[1] + "/pull/" + path_parts[3] +
```
Isn't `/pull/` already in `path_parts[2]`? Also please use `base::JoinString`.

bbondy replied:
More specifically `pull` but I agree `base::JoinString` will be a good cleanup here.

bbondy commented:
> Q: does this work with PR in private repos?

Yes. It does so intentionally.

bbondy commented:
I just squashed but here are the highlights:

- Added `kMaxBoundedStringDownloadSize` instead of using a magic number
- Used `JoinString` instead of string concat
- Removed explicit set of `request->credentials_mode`
- Renamed header file param to `url_loader_factory_for_test`

darkdh commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -348,11 +424,34 @@ std::string ExtractPdfContent(const ui::AXNode* pdf_root) {
+      base::JoinString({url.GetWithEmptyPath().spec(), path_parts[0],
+                        path_parts[1], path_parts[2], path_parts[3]},
+                       "/");
+  return GURL(base::JoinString({base_pr_url, "patch"}, "."));
```
This should be `base::StrAppend`

darkdh reviewed (approved):
++

diracdeltas commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -348,11 +424,35 @@ std::string ExtractPdfContent(const ui::AXNode* pdf_root) {
 
+// Obtains a patch URL from a pull request URL
+std::optional<GURL> GetGithubPatchURLForPRURL(const GURL& url) {
+  if (!url.is_valid() || !url.SchemeIsCryptographic() ||
```
would be better to check if scheme is exactly `https` unless we want to match other schemes.

bbondy replied:
Fixed in https://github.com/brave/brave-core/pull/22289/commits/60cbe140539a6a15af691030a9828e2be89846d9

diracdeltas commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -104,6 +128,33 @@ class PageContentFetcher {
         std::move(callback), invalidation_token));
   }
 
+  void StartGithub(
```
does this fail if the site (github) throws a certificate error? i would assume so.

diracdeltas commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -404,18 +505,23 @@ void FetchPageContent(content::WebContents* web_contents,
     }
   }
 #endif
+  auto* fetcher = new PageContentFetcher();
```
should fetcher/loader only be created if patch_url is nonempty?

bbondy replied:
It should be created even if it is empty because it will be used in the below block if not there:
>   mojo::Remote<mojom::PageContentExtractor> extractor;
> // GetRemoteInterfaces() cannot be null if the render frame is created.
> primary_rfh->GetRemoteInterfaces()->GetInterface(
      extractor.BindNewPipeAndPassReceiver());
> fetcher->Start(std::move(extractor), invalidation_token, loader,
                 std::move(callback));

ShivanKaul commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -390,8 +490,9 @@ void FetchPageContent(content::WebContents* web_contents,
     return;
   }
 
+  auto url = web_contents->GetURL();
```
There's a [AIChatTabHelper::GetPageURL](https://github.com/brave/brave-core/blob/master/components/ai_chat/content/browser/ai_chat_tab_helper.cc#L190) method which uses `web_contents()->GetLastCommittedURL()`

bbondy replied:
I just moved that usage but good catch, I'll fix.

bbondy commented:
> does this fail if the site (github) throws a certificate error? i would assume so.

Tried by advancing my clock and also by writing a small python server and setting my hosts file to 127.0.0.1 forgithub.com. But it doesn't let me bypass the interstitial in either case.   If you have any other ideas on how to test this pls lmk

petemill commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -267,7 +317,31 @@ class PageContentFetcher {
+               << loader->GetFinalURL().spec() << " status: " << response_code;
+    }
+    DVLOG(2) << "Got patch: " << patch_content;
+    SendResultAndDeleteSelf(std::move(callback), patch_content, "", false);
```
By not passing an invalidation token, I think we're performing these extra fetches unnecessarily every time the user submits a new AI Chat conversation entry. Perhaps we're ok with that because of network caching? If not, and we don't expect the content of the patch url to change during a conversation (i.e. during a Tab's page lifecycle), then we can use the patch url as the invalidation token. Then, if the next call to FetchContent provides the same patch url as is generated from `GetGithubPatchURLForPRURL`, the callback can be called immediately passing the same invalidation token and the caller (ConversationDriver) will know to use its content cache instead.
This is what's done for YouTube to avoid any unnecessary re-fetch and parsing during a conversation, and the plumbing is all setup.

bbondy replied:
Thanks I'll fix that in this follow up https://github.com/brave/brave-browser/issues/36442
